### PR TITLE
[Twig] CVA - Rework implementation + documentation

### PR DESCRIPTION
**Pull request by smnandre**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | ..
| License       | MIT

Try another implementation + add some tests.

Updated/added a lot of documentation in the CVA class.

All tests are green but i'm not 100% sure this made no change at all :| 

@WebMamba  I let you decide what you want to do with it (i needed a class to test some things :) ) .. really open to any choice :) 

@kbond that does not include your modifications i think :|

### CODE REVIEW

File: src/TwigComponent/src/CVA.php [hunk 1]
```diff
@@ -12,107 +12,134 @@
...
     }
 
-    public function apply(array $recipes, ?string ...$classes): string
+    /**
+     * @deprecated since Symfony 2.17. Use {@see apply()} instead.
+     */
+    public function resolve(array $recipes): string
     {
-        return trim($this->resolve($recipes).' '.implode(' ', array_filter($classes)));
+        trigger_deprecation('symfony/ux-twig-component', '2.17', 'The method "%s()" is deprecated and will be remove in 3.0. Use "apply()" instead.', __METHOD__);
```
Comment (kbond) [inline src/TwigComponent/src/CVA.php]: I think we should just kill this method (and not worry about deprecating). I doubt anyone is using it.

Comment (smnandre) [reply src/TwigComponent/src/CVA.php]: Good & thanks!

File: src/TwigComponent/src/Twig/ComponentExtension.php [hunk 2]
```diff
@@ -87,25 +87,29 @@ public function finishEmbeddedComponentRender(): void
...
+     * } $cva
      */
     public function cva(array $cva): CVA
     {
         return new CVA(
-            $cva['base'] ?? null,
-            $cva['variants'] ?? null,
-            $cva['compoundVariants'] ?? null,
-            $cva['defaultVariants'] ?? null,
+            $cva['base'] ?? [],
```
Comment (nicolas-grekas) [inline src/TwigComponent/src/Twig/ComponentExtension.php]: `''`?

Comment (smnandre) [reply src/TwigComponent/src/Twig/ComponentExtension.php]: Yep thanks!

File: src/TwigComponent/src/CVA.php [hunk 3]
```diff
@@ -12,107 +12,124 @@
...
  */
 final class CVA
 {
     /**
-     * @var string|list<string|null>|null
-     * @var array<string, array<string, string|list<string|null>>|null the array should have the following format [variantCategory => [variantName => classes]]
-     *                                                ex: ['colors' => ['primary' => 'bleu-8000', 'danger' => 'red-800 text-bold'], 'size' => [...]]
-     * @var array<array<string, string|array<string>>> the array should have the following format ['variantsCategory' => ['variantName', 'variantName'], 'class' => 'text-red-500']
-     * @var array<string, string>|null
+     * @var list<string>
```
Comment (WebMamba) [inline src/TwigComponent/src/CVA.php]: Are you sure list<string> include null values? Maybe list<string|null>

**Review by WebMamba — APPROVED** [review]
Thanks @smnandre Looks go to me! Thanks for the rework!

File: src/TwigComponent/src/CVA.php (hunk 3 above)
Comment (smnandre) [reply src/TwigComponent/src/CVA.php]: What is the use case to pass null values ?

Comment (WebMamba) [reply src/TwigComponent/src/CVA.php]: https://github.com/symfony/ux/issues/1570

Comment (smnandre) [reply src/TwigComponent/src/CVA.php]: Could we filter those values in the Twig function and/or the constructor ?

Comment (WebMamba) [reply src/TwigComponent/src/CVA.php]: Actually we test for null so I guess your code already works with nullable 😉

Comment (smnandre) [reply src/TwigComponent/src/CVA.php]: Ok so just update the docblock, right ?

Comment (smnandre) [reply src/TwigComponent/src/CVA.php]: Updated!

**Review by kbond — APPROVED** [review]

**Comment by kbond** [comment]
Thanks Simon.

**Pull request merged**